File: src/main/java/com/qtech/bubbles/core/utils/categories/GraphicsUtils.java
package com.qtech.bubbles.core.utils.categories;

import com.qtech.bubbles.common.GraphicsProcessor;

import java.awt.*;
import java.awt.geom.Point2D;
import java.awt.geom.Rectangle2D;

public class GraphicsUtils {
    /**
     * Draw a String centered in the middle of a Rectangle.
     *
     * @param g    The GraphicsProcessor instance.
     * @param text The String to draw.
     * @param rect The Rectangle to center the text within.
     * @param font The Font for the text.
     * @author <b>Danier Kvist</b> <a href="https://stackoverflow.com/a/27740330/11124294">from this answer</a>.
     */
    public static void drawCenteredString(GraphicsProcessor g, String text, Rectangle2D rect, Font font) {
        // Get the FontMetrics
        FontMetrics metrics = g.getFontMetrics(font);

        // Determine the X coordinate for the text
        int x = (int) (rect.getX() + (rect.getWidth() - metrics.stringWidth(text)) / 2);

        // Determine the Y coordinate for the text (note we add the ascent, as in java 2d 0 is top of the screen)
        int y = (int) (rect.getY() + ((rect.getHeight() - metrics.getHeight()) / 2) + metrics.getAscent());

        // Set the font
        g.setFont(font);

        // Draw the String
        g.drawString(text, x, y);
    }

    /**
     * Draw a String centered in the middle of a Rectangle.
     *
     * @param g      The GraphicsProcessor instance.
     * @param text   The String to draw.
     * @param point  The Point to center the text within.
     * @param height The height to center the text within.
     * @param font   The Font for the text.
     * @author <b>Danier Kvist</b> <a href="https://stackoverflow.com/a/27740330/11124294">from this answer</a>.
     */
    public static void drawRightAnchoredString(GraphicsProcessor g, String text, Point2D point, double height, Font font) {
        // Get the FontMetrics
        FontMetrics metrics = g.getFontMetrics(font);

        // Determine the X coordinate for the text
        int x = (int) (point.getX() - (metrics.stringWidth(text)));

        // Determine the Y coordinate for the text (note we add the ascent, as in java 2d 0 is top of the screen)
        int y = (int) (point.getY() + ((height - metrics.getHeight()) / 2) + metrics.getAscent());

        // Set the font
        g.setFont(font);

        // Draw the String
        g.drawString(text, x, y);
    }

    /**
     * Draw a String centered in the middle of a Rectangle.
     *
     * @param g      The GraphicsProcessor instance.
     * @param text   The String to draw.
     * @param point  The Point to center the text within.
     * @param height The height to center the text within.
     * @param font   The Font for the text.
     * @author <b>Danier Kvist</b> <a href="https://stackoverflow.com/a/27740330/11124294">from this answer</a>.
     */
    public static void drawLeftAnchoredString(GraphicsProcessor g, String text, Point2D point, double height, Font font) {
        // Get the FontMetrics
        FontMetrics metrics = g.getFontMetrics(font);

        // Determine the X coordinate for the text
        int x = (int) (point.getX());

        // Determine the Y coordinate for the text (note we add the ascent, as in java 2d 0 is top of the screen)
        int y = (int) (point.getY() + ((height - metrics.getHeight()) / 2) + metrics.getAscent());

        // Set the font
        g.setFont(font);

        // Draw the String
        g.drawString(text, x, y);
    }
}

File: src/main/kotlin/com/qtech/bubbles/core/utils/categories/GraphicsUtils.kt
package com.qtech.bubbles.core.utils.categories

import com.qtech.bubbles.graphics.GraphicsProcessor
import java.awt.Font
import java.awt.geom.Point2D
import java.awt.geom.Rectangle2D

object GraphicsUtils {
    /**
     * Draw a String centered in the middle of a Rectangle.
     *
     * @param g    The GraphicsProcessor instance.
     * @param text The String to draw.
     * @param rect The Rectangle to center the text within.
     * @param font The Font for the text.
     * @author **Danier Kvist** [from this answer](https://stackoverflow.com/a/27740330/11124294).
     */
    @JvmStatic
    fun drawCenteredString(g: GraphicsProcessor, text: String, rect: Rectangle2D, font: Font?) {
        // Get the FontMetrics
        val metrics = g.fontMetrics(font)

        // Determine the X coordinate for the text
        val x = (rect.x + (rect.width - metrics.stringWidth(text)) / 2).toInt()

        // Determine the Y coordinate for the text (note we add the ascent, as in java 2d 0 is top of the screen)
        val y = (rect.y + (rect.height - metrics.height) / 2 + metrics.ascent).toInt()

        // Set the font
        g.font(font)

        // Draw the String
        g.textI(text, x, y)
    }

    /**
     * Draw a String centered in the middle of a Rectangle.
     *
     * @param g      The GraphicsProcessor instance.
     * @param text   The String to draw.
     * @param point  The Point to center the text within.
     * @param height The height to center the text within.
     * @param font   The Font for the text.
     * @author **Danier Kvist** [from this answer](https://stackoverflow.com/a/27740330/11124294).
     */
    @JvmStatic
    fun drawRightAnchoredString(g: GraphicsProcessor, text: String, point: Point2D, height: Double, font: Font?) {
        // Get the FontMetrics
        val metrics = g.fontMetrics(font)

        // Determine the X coordinate for the text
        val x = (point.x - metrics.stringWidth(text)).toInt()

        // Determine the Y coordinate for the text (note we add the ascent, as in java 2d 0 is top of the screen)
        val y = (point.y + (height - metrics.height) / 2 + metrics.ascent).toInt()

        // Set the font
        g.font(font)

        // Draw the String
        g.textI(text, x, y)
    }

    /**
     * Draw a String centered in the middle of a Rectangle.
     *
     * @param g      The GraphicsProcessor instance.
     * @param text   The String to draw.
     * @param point  The Point to center the text within.
     * @param height The height to center the text within.
     * @param font   The Font for the text.
     * @author **Danier Kvist** [from this answer](https://stackoverflow.com/a/27740330/11124294).
     */
    @JvmStatic
    fun drawLeftAnchoredString(g: GraphicsProcessor, text: String?, point: Point2D, height: Double, font: Font?) {
        // Get the FontMetrics
        val metrics = g.fontMetrics(font)

        // Determine the X coordinate for the text
        val x = point.x.toInt()

        // Determine the Y coordinate for the text (note we add the ascent, as in java 2d 0 is top of the screen)
        val y = (point.y + (height - metrics.height) / 2 + metrics.ascent).toInt()

        // Set the font
        g.font(font)

        // Draw the String
        g.textI(text, x, y)
    }
}